src/boot/efi/chid.c

```diff
@@ -0,0 +1,132 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+
+/*
+ * Based on Nikita Travkin's dtbloader implementation.
+ * Copyright (c) 2024 Nikita Travkin <nikita@trvn.ru>
+ *
+ * https://github.com/TravMurav/dtbloader/blob/main/src/chid.c
+ */
+
+/*
+ * Based on Linaro dtbloader implementation.
+ * Copyright (c) 2019, Linaro. All rights reserved.
+ *
+ * https://github.com/aarch64-laptops/edk2/blob/dtbloader-app/EmbeddedPkg/Application/ConfigTableLoader/CHID.c
+ */
+
+#include "chid.h"
+#include "chid-fundamental.h"
+#include "efi.h"
+#include "sha1-fundamental.h"
+#include "smbios.h"
+#include "util.h"
+
+/**
+ * smbios_to_hashable_string() - Convert ascii smbios string to stripped char16_t.
+ */
+static char16_t *smbios_to_hashable_string(const char *str) {
+        char16_t *ret;
+
+        if (!str) {
+                /* User of this function is expected to free the result. */
+                ret = xnew(char16_t, 1);
+                return ret;
```

Comment:
hmm, this memory will not be initialized.

please add:

```c
#define xnew0(type, n) ((type *) xcalloc_multiply((n), sizeof(type)))
```

to util.h and use that.

(we also have a `new0()` macro in userspace, hence this would be nicely systematic)